### PR TITLE
Dendrite hyperparameter search

### PR DESCRIPTION
**Pull request by jeremyforest**
- `Makefile` allows to run all analysis and figure scripts from the data (I have not added the steps to run the config for generating the searches)
- modified `analyze_results.py` to consider either best, last task or all results
- `hyperparameter_search.py` contains all configs to run the searches
- `hyperparameter_figures.py` contains all code to make the figures stored in figs. 

Ref Jira #RES-2226 ,  #RES-2227, #RES-2250

### CODE REVIEW

File: projects/dendrites/permutedMNIST/experiments/__init__.py [hunk 1]
```diff
@@ -36,6 +38,10 @@
 # Collect all configurations
 CONFIGS = dict()
 CONFIGS.update(BASE)
+CONFIGS.update(CENTROID)
+CONFIGS.update(HYPERPARAMETERSEARCH)
+CONFIGS.update(SP_PROTO)
+CONFIGS.update(SP_CONTEXT)
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/__init__.py:44]: Here, you only need `CONFIGS.update(HYPERPARAMETERSEARCH)` and should remove the other 3 lines, because they already exist in this file!

File: projects/dendrites/permutedMNIST/experiments/base.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ class PermutedMNISTExperiment(mixins.RezeroWeights,
         root=os.path.expanduser("~/nta/results/data/"),
         dim_context=1024,
         seed=42,
-        download=False,  # Change to True if running for the first time
+        download=True,  # Change to True if running for the first time
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/base.py:63]: Can you revert this to `download=False`?

File: projects/dendrites/permutedMNIST/experiments/cns2021_figure1c_data/kw_sparsity_search.csv [hunk 3]
```diff
@@ -0,0 +1,81 @@
+,Experiment name,Activation sparsity,FF weight sparsity,Dendrite weight sparsity,Num segments,Dim context,Epochs,Num tasks,LR,Momentum,Seed,Accuracy,ID
+0,sparsity_search,0.01,0.5,0.0,10,784,3,10,0.0005,0.0,2772,0.9366399645805359,sparsity_search kw_percent_on=0.01
+1,sparsity_search,0.01,0.5,0.0,10,784,3,10,0.0005,0.0,7630,0.9363999962806702,sparsity_search kw_percent_on=0.01
+2,sparsity_search,0.01,0.5,0.0,10,784,3,10,0.0005,0.0,9705,0.9332499504089355,sparsity_search kw_percent_on=0.01
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/cns2021_figure1c_data/kw_sparsity_search.csv:4]: I don't think we need to add any CSV files, since the code itself can generate all CSV output.

Comment (karangrewal) [reply projects/dendrites/permutedMNIST/experiments/cns2021_figure1c_data/kw_sparsity_search.csv:4]: This applies to pkl files too.

File: projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+#
+#  You should have received a copy of the GNU Affero Public License
+#  along with this program.  If not, see htt"://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+#
+
+"""
+Search hyperparameter configuration
+"""
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py:23]: It would be really helpful if you provide more detail in the file docstring.

File: projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+
+if __name__ == "__main__":
+
+    savefigs = True
+    figs_dir = "figs/"
+    if savefigs:
+        if not os.path.isdir(f"{figs_dir}"):
+            os.makedirs(f"{figs_dir}")
+
+    experiment_folder = "~/nta/nupic.research/projects/dendrites/permutedMNIST/experiments/data_hyperparameter_search/"
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py:382]: You shouldn't have hardcoded paths like this in any code that you check in!

**Review by karangrewal — CHANGES_REQUESTED** [review]
TLDR; you should remove all csv files, they don't need to be added to the repository

**Review by subutai — COMMENTED** [review]
One high level comment is that the `experiments` directory is only for Vernon experiment files. Can you create a separate subdirectory under `permutedMNIST` for the analysis files?

File: projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py (hunk 5 above)
Comment (subutai) [reply projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py:382]: You can make it relative to the location of the script, such as `../experiments/...`

File: projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+#  Numenta Platform for Intelligent Computing (NuPIC)
```
Comment (subutai) [inline projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py:1]: This file should stay under `experiments` since its a Vernon experiment config file.

File: projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py (hunk 5 above)
Comment (jeremyforest) [reply projects/dendrites/permutedMNIST/experiments/hyperparameters_figures.py]: done

File: projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py (hunk 6 above)
Comment (jeremyforest) [reply projects/dendrites/permutedMNIST/experiments/hyperparameter_search.py:1]: should be done if I understood correctly what you meant.

File: projects/dendrites/permutedMNIST/analyze_results.py [hunk 7]
```diff
@@ -41,7 +41,42 @@ def key_func(x):
...
+    num_tasks = config["num_tasks"]
+    lr = config["optimizer_args"]["lr"]
+    momentum = config["optimizer_args"].get("momentum", 0.0)
+    iteration = results["training_iteration"]
+
+    # This list must match the column headers in collect_results
+    df_entries.append(
+        [
+            exp,
+            kw_percent_on,
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/analyze_results.py:62]: Here, you modify `df_entries` but don't return anything. Are you sure that it stays modified after the function call ends?

File: projects/dendrites/permutedMNIST/analyze_results.py [hunk 8]
```diff
@@ -72,47 +107,67 @@ def parse_one_experiment(exp, state, df):
...
+                        print("using parsing method : best")
+                        best_results = max(
+                            results, key=lambda x: x.get("mean_accuracy", 0.0)
+                        )
+                        best_result = best_results["mean_accuracy"]
+                        if best_result > 0.0:
+                            parse(
+                                best_result,
+                                results,
+                                trial_checkpoint,
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/analyze_results.py:122]: For better aesthetics, I recommend not having 1 argument per line, but instead all on the same line and wrapping, e.g.

```
if best_result > 0.0:
    parse(argument_1, argument_2, argument_3,
          argument_4, argument_5)
```

Comment (karangrewal) [reply projects/dendrites/permutedMNIST/analyze_results.py:122]: The above comment applies to 3 different calls to `parse`

File: projects/dendrites/permutedMNIST/experiments/__init__.py [hunk 9]
```diff
@@ -20,13 +20,12 @@
 # ----------------------------------------------------------------------
 
 from .base import CONFIGS as BASE
+from .hyperparameter_search import CONFIGS as HYPERPARAMETERSEARCH
 from .batch import CONFIGS as BATCH
 from .batch_mnist import CONFIGS as BATCH_MNIST
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/__init__.py:25]: This init file has kept all imports in alphabetical order, please keep it that way!

File: projects/dendrites/permutedMNIST/experiments/__init__.py [hunk 10]
```diff
@@ -36,10 +35,10 @@
 # Collect all configurations
 CONFIGS = dict()
 CONFIGS.update(BASE)
+CONFIGS.update(CENTROID)
+CONFIGS.update(HYPERPARAMETERSEARCH)
 CONFIGS.update(BATCH)
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/__init__.py:40]: 1. note alphabetical order,
2. the line `CONFIGS.update(CENTROID)` is now repeating (a few lines later), so I don't think you need to add it here.

File: projects/dendrites/permutedMNIST/experiments/__init__.py [hunk 11]
```diff
@@ -20,13 +20,12 @@
...
 
 from .base import CONFIGS as BASE
+from .hyperparameter_search import CONFIGS as HYPERPARAMETERSEARCH
 from .batch import CONFIGS as BATCH
 from .batch_mnist import CONFIGS as BATCH_MNIST
 from .centroid import CONFIGS as CENTROID
 from .no_dendrites import CONFIGS as NO_DENDRITES
 from .si_centroid import CONFIGS as SI_CENTROID
-from .sp_context import CONFIGS as SP_CONTEXT
-from .sp_context_search import CONFIGS as SP_PROTO
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/__init__.py:29]: why delete these?

File: projects/dendrites/permutedMNIST/experiments/base.py [hunk 12]
```diff
@@ -140,19 +135,19 @@ class PermutedMNISTExperiment(mixins.RezeroWeights,
...
         dendrite_weight_sparsity=0.0,
         weight_sparsity=tune.sample_from(
-            lambda spec: np.random.choice([0.90, 0.8, 0.7, 0.5, 0.25, 0.0])),
+            lambda spec: np.random.choice([0.90, 0.8, 0.7, 0.5, 0.25, 0.0])
+        ),
     ),
-
     epochs=tune.sample_from(lambda spec: np.random.randint(1, 3)),
     num_samples=5,
-
```
Comment (karangrewal) [inline projects/dendrites/permutedMNIST/experiments/base.py:155]: Why undo the blank line removal? They help for readability.

File: projects/dendrites/permutedMNIST/experiments/__init__.py (hunk 9 above)
Comment (jeremyforest) [reply projects/dendrites/permutedMNIST/experiments/__init__.py:25]: solved with Isort

File: projects/dendrites/permutedMNIST/experiments/base.py (hunk 12 above)
Comment (jeremyforest) [reply projects/dendrites/permutedMNIST/experiments/base.py:155]: automatic formatting using black and isort.

**Review by karangrewal — APPROVED** [review]
